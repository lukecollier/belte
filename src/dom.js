import parse5 from 'parse5';
import { reduce, assoc, filter, clone, not, pick, omit, isNil, partial, mapObjIndexed, toPairs, is, pipe } from 'ramda';
import * as dom5 from 'dom5';

const HTML5_ELEMENT_NAMES = ['html', 'div', 'base', 'head', 'link', 'meta', 'style', 'title', 'body', 'address', 'article', 'aside', 'footer', 'header', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hgroup', 'main', 'nav', 'section', 'blockquote', 'dd', 'dir', 'dl', 'dt', 'figcaption', 'figure', 'hr', 'li', 'main', 'ol', 'p', 'pre', 'ul', 'a', 'abbr', 'b', 'bdi', 'bdo', 'br', 'cite', 'code', 'data', 'dfn', 'em', 'i', 'kbd', 'mark', 'q', 'rb', 'rtc', 'ruby', 's', 'samp', 'small', 'span', 'strong', 'sub', 'sup','time', 'u', 'var', 'wbr', 'area', 'audio', 'img', 'map', 'track', 'video', 'applet', 'embed', 'iframe', 'noembed', 'object', 'param', 'picture', 'source', 'canvas', 'noscript', 'script', 'del', 'ins', 'caption', 'col', 'colgroup', 'table', 'tbody', 'td', 'tfoot', 'th', 'thead', 'tr', 'button', 'datalist', 'fieldset', 'form', 'input', 'label', 'legend', 'meter', 'optgroup', 'option', 'output', 'progress', 'select', 'textarea', 'details', 'dialog', 'menu', 'menuitem', 'summary', 'content', 'element', 'shadow', 'slot', 'template'];
export const isCustomElement = (tagNode) => {
  return !HTML5_ELEMENT_NAMES.includes(tagNode.tagName);
}

export const parse = (html) => parse5.parse(html, {scriptingEnabled:false});

export const serialize = (dom) => parse5.serialize(dom);

export const appendToHead = (dom, headElements) => {
  const rendered = clone(dom);
  const headSet = new Set(headElements);
  const head = dom5.query(dom, (el) => el.tagName === 'head');
  headSet.forEach(elementHtml => 
    dom5.append(head, parse5.parseFragment(elementHtml, {scriptingEnabled:false}))
  );
  return rendered;
}

const serializeAttribs = 
  pipe(toPairs, reduce((acc, [name, value]) => `${acc}${name}="${value}"`,''));

const id = (num) => `belte-component-${num}`; 

const assocNameValue = (acc, {name, value}) => assoc(name, value, acc);

export const replaceWith = (dom, render, resolveSrc) => {
  const rendered = clone(dom);
  const reservedAttrs = ['class', 'style', 'href']
  const body = dom5.query(rendered, (el) => el.tagName === 'body');
  dom5.queryAll(body, (el) => isCustomElement(el)).forEach((el, i) => {
    const attrs = reduce(assocNameValue, {}, el.attrs);
    const htmlAttrs = pick(reservedAttrs, attrs);
    const props = omit(reservedAttrs, attrs);
    const newEl = parse5.parseFragment(`
      <div id="${id(i)}" ${serializeAttribs(htmlAttrs)}>
        ${render(resolveSrc[el.tagName], props)}
      </div>`, { scriptingEnabled: false });
    dom5.replace(el, newEl);
  });
  return rendered;
}

export const customElements = (dom) => {
  const body = dom5.query(dom, (el) => el.tagName === 'body');
  return dom5.queryAll(body, (el) => isCustomElement(el)).map((el, i) => {
    const attribs = reduce(assocNameValue, {}, el.attrs);
    return {id: id(i), attr: attribs, name: el.tagName}
  });
}
