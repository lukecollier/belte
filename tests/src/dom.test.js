import { customElements, parse, replaceWith, serialize } from '../../src/dom.js';

test('an empty document returns no custom element customElements', () => {
  const parsed = parse("<!DOCTYPE html><html><head><title></title></head><body></body></html>");
  const result = customElements(parsed);
  expect(result).toEqual([]);
});

test('an document with elements returns customElements', () => {
  const parsed = parse("<!DOCTYPE html><html><head><title></title></head><body><custom-element></custom-element></body></html>");
  const result = customElements(parsed);
  expect(result).toEqual([{id: 'belte-component-0', attr: {}, name: 'custom-element'}]);
});

test('an document with elements returns customElements with attribs', () => {
  const parsed = parse(`<!DOCTYPE html><html><head><title></title></head><body><custom-element test="hello"></custom-element></body></html>`);
  const result = customElements(parsed);
  expect(result).toEqual([{id: 'belte-component-0', attr: {test: "hello"}, name: 'custom-element'}]);
});

test('an document with elements returns customElements with multiple elements', () => {
  const parsed = parse(`<!DOCTYPE html><html><head><title></title></head><body><custom-element></custom-element><custom-element-two></custom-element-two></body></html>`);
  const result = customElements(parsed);
  expect(result).toEqual([{id: 'belte-component-0', attr: {}, name: 'custom-element'}, {id: 'belte-component-1', attr: {}, name: 'custom-element-two'}]);
});

test('an document with elements returns customElements with multiple elements and attribs', () => {
  const parsed = parse(`<!DOCTYPE html><html><head><title></title></head><body><custom-element number="one"></custom-element><custom-element-two number="two"></custom-element-two></body></html>`);
  const result = customElements(parsed);
  expect(result).toEqual([{id: 'belte-component-0', attr: {number: "one"}, name: 'custom-element'}, {id: 'belte-component-1', attr: {number: "two"}, name: 'custom-element-two'}]);
});

test('can replace custom element', () => {
  const parsed = parse(`<!DOCTYPE html><html><head><title></title></head><body><custom-element></custom-element></body></html>`);
  const result = replaceWith(parsed, (name, _) => name, (name) => name + '-resolved');
  expect(serialize(result)).toMatchSnapshot();
});

test('can replace custom multiple elements', () => {
  const parsed = parse(`<!DOCTYPE html><html><head><title></title></head><body><custom-element></custom-element><custom-element-two></custom-element-two></body></html>`);
  const result = replaceWith(parsed, (name, _) => name, (name) => name + '-resolved');
  expect(serialize(result)).toMatchSnapshot();
});

test('can replace custom element with attributes', () => {
  const parsed = parse(`<!DOCTYPE html><html><head><title></title></head><body><custom-element name="jeff"></custom-element></body></html>`);
  const result = replaceWith(parsed, (name, attr) => name + '-with-' + JSON.stringify(attr), (name) => name + '-resolved');
  expect(serialize(result)).toMatchSnapshot();
});

test('can replace custom element with multiple attributes', () => {
  const parsed = parse(`<!DOCTYPE html><html><head><title></title></head><body><custom-element name="jeff" second="mcgee"></custom-element></body></html>`);
  const result = replaceWith(parsed, (name, attr) => name + '-with-' + JSON.stringify(attr), (name) => name + '-resolved');
  expect(serialize(result)).toMatchSnapshot();
});

test('can replace custom element with multiple html attributes', () => {
  const parsed = parse(`<!DOCTYPE html><html><head><title></title></head><body><custom-element name="jeff" second="mcgee" class="nonsense"></custom-element></body></html>`);
  const result = replaceWith(parsed, (name, attr) => name + '-with-' + JSON.stringify(attr), (name) => name + '-resolved');
  expect(serialize(result)).toMatchSnapshot();
});
