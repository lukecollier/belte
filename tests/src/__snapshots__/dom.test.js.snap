// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can replace custom element 1`] = `
"<!DOCTYPE html><html><head><title></title></head><body>
      <div id=\\"belte-component-0\\">
        custom-element-resolved
      </div></body></html>"
`;

exports[`can replace custom element with attributes 1`] = `
"<!DOCTYPE html><html><head><title></title></head><body>
      <div id=\\"belte-component-0\\">
        custom-element-resolved-with-{\\"name\\":\\"jeff\\"}
      </div></body></html>"
`;

exports[`can replace custom element with multiple attributes 1`] = `
"<!DOCTYPE html><html><head><title></title></head><body>
      <div id=\\"belte-component-0\\">
        custom-element-resolved-with-{\\"name\\":\\"jeff\\",\\"second\\":\\"mcgee\\"}
      </div></body></html>"
`;

exports[`can replace custom element with multiple html attributes 1`] = `
"<!DOCTYPE html><html><head><title></title></head><body>
      <div id=\\"belte-component-0\\" class=\\"nonsense\\">
        custom-element-resolved-with-{\\"name\\":\\"jeff\\",\\"second\\":\\"mcgee\\"}
      </div></body></html>"
`;

exports[`can replace custom multiple elements 1`] = `
"<!DOCTYPE html><html><head><title></title></head><body>
      <div id=\\"belte-component-0\\">
        custom-element-resolved
      </div>
      <div id=\\"belte-component-1\\">
        custom-element-two-resolved
      </div></body></html>"
`;
